docs/contributor_guide/code_quality.md

```diff
@@ -0,0 +1,48 @@
+# Code quality
+
+```{seealso}
+Complete [dev install](./development_setup) instructions before continuing.
+```
+
+We have several tools configured for checking code quality:
+
+- Pre-commit checks run automatically at commit time.
+  Install checks with `pre-commit install`.
+  Run them manually with `pre-commit run --all-files`.
+  **Will exit non-zero when finding errors or changing files.**
+
+  - Ruff formats and lints (sometimes autofixes) Python code.
+
+  - Generic pre-commit checks help avoid common mistakes like committing large
+    files or trailing whitespace.
+
+- Package scripts (defined in `package.json`) to check (and/or fix)
+  TypeScript, JavaScript, CSS, JSON, Markdown, and YAML.
+  Run manually with `jlpm run lint`.
+  **Will exit 0 when applying fixes.
+  Check the logs and/or `git status` after every run.**
+
+  - Prettier formats the file types listed above.
+
+  - Eslint lints (sometimes autofixes) JS/TS code.
+
+- UI tests using [Galata](https://github.com/jupyterlab/galata), defined in the
```

Comment:
I would put a big warning here. Running the tests locally isn't reliable and I'd not recommend it, things like font rendering and WebGL depend a lot on the browser and GPU you're using.

So I'd recommend not taking the local runs for granted, and not updating the snapshots from a local run.

I always wondered if tools like [`act`](https://github.com/nektos/act) made things better locally, though I never tried really, and I guess all GPU related things are still dependent on the running machine.

Reply:
>things like font rendering and WebGL depend a lot on the browser and GPU you're using

The GPU, too!? Wow. I thought the browser thing was handled by playwright downloading the same version of Chromium for every test run. Oof. Local runs _without_ snapshot tests should be OK, though, right? Maybe there's a way we can exclude them and document that for local running. Having the Playwright Inspector running locally really helped with debugging the test.

Reply:
5c97cd0 How's that?

Reply:
Oh neat! I did not know you could tell playwright to ignore snapshots!